python_modules/libraries/dagster-sdf/dagster_sdf/sdf_information_schema.py

```diff
@@ -142,6 +154,60 @@ def build_sdf_multi_asset_args(
 
         return deps, outs, internal_asset_deps
 
+    def get_columns(self) -> Dict[str, List[TableColumn]]:
+        columns = self.read_table("columns")[
+            ["table_id", "column_id", "classifiers", "column_name", "datatype", "description"]
+        ]
+        table_columns: Dict[str, List[TableColumn]] = {}
+        for row in columns.rows(named=True):
+            if row["table_id"] not in table_columns:
+                table_columns[row["table_id"]] = []
+            table_columns[row["table_id"]].append(
+                TableColumn(
+                    name=row["column_name"],
+                    type=row["datatype"],
+                    description=row["description"],
+                    constraints=TableColumnConstraints(other=row["classifiers"]),
+                )
+            )
+        return table_columns
+
+    def stream_asset_observations(
+        self, dagster_sdf_translator: DagsterSdfTranslator
+    ) -> Iterator[SdfDagsterEventType]:
+        table_columns = self.get_columns()
+        tables = self.read_table("tables").filter(
+            ~pl.col("purpose").is_in(["system", "external-system"])
+        )
+        for table_row in tables.rows(named=True):
+            asset_key = dagster_sdf_translator.get_asset_key(table_row["table_id"])
+            code_references = None
+            for source_location in table_row["source_locations"]:
+                if source_location.endswith(".sql"):
+                    code_references = CodeReferencesMetadataSet(
+                        code_references=CodeReferencesMetadataValue(
+                            code_references=[
+                                LocalFileCodeReference(
+                                    file_path=os.fspath(
+                                        self.workspace_dir.joinpath(source_location)
+                                    )
+                                )
+                            ]
+                        )
+                    )
+            metadata = {
+                **TableMetadataSet(
+                    column_schema=TableSchema(
+                        columns=table_columns.get(table_row["table_id"], []),
+                    ),
+                    relation_identifier=table_row["table_id"],
+                ),
+                **(code_references if code_references else {}),
+            }
```

Comment:
This is metadata that can be inferred at definition time, rather than run time. We shouldn't need to emit `AssetObservation`'s for this.

Can we add this information to the definitional `AssetOut`'s instead?

Reply:
This is partially true if the upstreams are alive at initialization time.  If we depend on an upstream table that is materialized via a dagster asset for example, there is no way to generate the schema unless we use schema inference, and even in this case there is no guarantee it will be accurate!

For now, I'm outputting these at materialization time, but we can do an "if available, surface at compile time" approach in the future

Reply:
If we need to emit this information at run time, why can't we add this information on the `AssetMaterialization` itself?

Reply:
I do agree that it should be on the `AssetMaterialization`. 👍